### PR TITLE
[MPoW] Dynamic used size according to the count of groups

### LINKED ISSUE #370: [GPoS] Stake limit of files should be dynamically changed according to the group number
Since we need to avoid the *local store attack* from node, we need to design a brand new mapping from `file_size` to `used`, so the whole flow from `file_size` to `stake_limit` shows below:

image

And the solution of `fn1` is:
1. Loop all `anchors` in `used_info`, check if work report reported
2. Query `fn1` to get ratio
3. Update all `group_owner`'s used

> `delete_file` may delayed update used size the  by `added_file`

### PR DESCRIPTION
**Pull request by mmyyrroonn**
Fix #370
Add two more todo to fix the bugs #409 #410

### CODE REVIEW

File: cstrml/market/src/lib.rs [hunk 1]
```diff
@@ -688,7 +700,7 @@ impl<T: Config> Module<T> {
             UsedTrashII::insert(cid, used_info.clone());
             UsedTrashSizeII::mutate(|value| {*value += 1;});
             // archive used for each merchant
-            for anchor in used_info.groups.iter() {
+            for (anchor, _) in used_info.groups.iter() {
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:703]: Change to foreach `keys`?

File: cstrml/market/src/lib.rs [hunk 2]
```diff
@@ -859,9 +872,13 @@ impl<T: Config> Module<T> {
         
         // 1. Delete files anchor
         <Files<T>>::mutate(cid, |maybe_f| match *maybe_f {
-            Some((_, ref mut used_info)) => {
-                if used_info.groups.take(anchor).is_some() {
-                    used_size = used_info.used_size;
+            Some((ref file_info, ref mut used_info)) => {
+                if let Some(is_calculated_as_reported_group_count) = used_info.groups.remove(anchor) {
+                    used_size = used_info.used_size; // need to delete the used_size before the update
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:877]: what if `is_calculated_as_reported_group_count ` = false?

File: cstrml/market/src/lib.rs [hunk 3]
```diff
@@ -925,8 +940,34 @@ impl<T: Config> Module<T> {
...
+    fn update_other_wr_used(groups: &BTreeMap<SworkerAnchor, bool>, prev_used_size: u64, new_used_size: u64) {
+        if prev_used_size == new_used_size {
+            return;
+        }
+        for (anchor, _) in groups.iter() {
+            T::SworkerInterface::update_used(anchor, prev_used_size, new_used_size);
+        }
+    }
+
+    fn count_valid_groups(groups: &mut BTreeMap<SworkerAnchor, bool>, curr_bn: BlockNumber) -> u32 {
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:958]: `count_reported_groups`

File: cstrml/market/src/lib.rs [hunk 4]
```diff
@@ -925,8 +940,34 @@ impl<T: Config> Module<T> {
...
+    }
+
+    fn update_other_wr_used(groups: &BTreeMap<SworkerAnchor, bool>, prev_used_size: u64, new_used_size: u64) {
+        if prev_used_size == new_used_size {
+            return;
+        }
+        for (anchor, _) in groups.iter() {
+            T::SworkerInterface::update_used(anchor, prev_used_size, new_used_size);
+        }
+    }
```
Comment (zikunfan) [inline cstrml/market/src/lib.rs:956]: merge with caller fn

**Review by badkk — DISMISSED** [review]

File: cstrml/market/src/lib.rs [hunk 5]
```diff
@@ -188,13 +190,14 @@ impl<T: Config> MarketInterface<<T as system::Config>::AccountId, BalanceOf<T>>
...
-            
+
             // 2. Delete replica from file_info
             if let Some((mut file_info, used_info)) = <Files<T>>::get(cid) {
                 // if this anchor didn't report work, we already decrease the `reported_replica_count` in `calculate_payout`
                 if T::SworkerInterface::is_wr_reported(&anchor, claimed_bn) {
                     // decrease it due to deletion
-                    file_info.reported_replica_count -= 1;
+                    // TODO: there is a bug
+                    file_info.reported_replica_count = file_info.reported_replica_count.saturating_sub(1);
```
[the remark this answers is not in the document]
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:200]: yes. Part of and will be fixed in the next PR.

File: cstrml/market/src/lib.rs (hunk 2 above)
Comment (MyronFanQiu) [reply cstrml/market/src/lib.rs:877]: we should still delete used no matter the value of `is_calculated_as_reported_group_count`. I added some comments here.

**Review by badkk — APPROVED** [review]

**Pull request merged**